src/wallet/rpcwallet.cpp

```diff
@@ -4766,8 +4766,12 @@ UniValue getpeginaddress(const JSONRPCRequest& request)
     pwallet->AddCScript(dest_script);
 
     // Get P2CH deposit address on mainchain from most recent fedpegscript.
-    const std::vector<CScript>& fedpegscripts = GetValidFedpegScripts(chainActive.Tip(), Params().GetConsensus(), true /* nextblock_validation */);
-    CTxDestination mainchain_dest(ScriptHash(GetScriptForWitness(calculate_contract(fedpegscripts.front(), dest_script))));
+    const auto& fedpegscripts = GetValidFedpegScripts(chainActive.Tip(), Params().GetConsensus(), true /* nextblock_validation */);
+    CTxDestination mainchain_dest(WitnessV0ScriptHash(calculate_contract(fedpegscripts.front().second, dest_script)));
+    // P2SH-wrapped is the only valid choice for non-dynafed chains
+    if (!IsDynaFedEnabled(chainActive.Tip(), Params().GetConsensus())) {
+        mainchain_dest = ScriptHash(GetScriptForDestination(mainchain_dest));
+    }
```

Comment:
We need a similar check for dynafed chains.

If dynafed is enabled then it is necessary to look at the `fedpeg_program` field; if this is p2sh-wrapped then we need to p2sh-wrap the mainchain address.

Reply:
fixed. Need to add a test for this.